### PR TITLE
Add new lint for byte char slices

### LINKED ISSUE #10147: Use byte string instead of slice of byte characters
### What it does

I accidentially used

```rust
let a = &[b'a', b'/', b'b'];
```

instead of

```rust
let a = b"a/b";
```

so I figured this could be a clippy lint.

### Lint Name

bytestring_instead_of_byte_chars

### Category

style

### Advantage

- Easier to read
- Easier to change/extend

### Drawbacks

None

### Example

```rust
let a = &[b'a', b'/', b'b'];
```

Could be written as:

```rust
let a = b"a/b";
```

### PR DESCRIPTION
**Pull request by TheNeikos**
This patch adds a new lint that checks for potentially harder to read byte char slices: `&[b'a', b'b']` and suggests to replace them with the easier to read `b"ab"` form.

Fixes #10147 

---

changelog: new lint: [`byte_char_slices`]
[#10155](https://github.com/rust-lang/rust-clippy/pull/10155)

### CODE REVIEW

**Comment by TheNeikos** [comment]
Potentially also @Alexendoo ? (Since you 'approved' the idea)

File: clippy_lints/src/byte_char_slice.rs
Comment (Alexendoo) [inline clippy_lints/src/byte_char_slice.rs:79]: We can collect into an `Option<String>` from an iterator of `Option<&str>`s, that'll let us skip doing the validation in `all_member_bytes` separately

```suggestion
                members
                    .iter()
                    .map(|member| match &member.kind {
                        ExprKind::Lit(Lit {
                            kind: LitKind::Byte,
                            symbol,
                            ..
                        }) => Some(symbol.as_str()),
                        _ => None,
                    })
                    .collect()
```

File: clippy_lints/src/byte_char_slice.rs
Comment (Manishearth) [inline clippy_lints/src/byte_char_slice.rs:74]: suggestion: should we be escaping these somehow, or retaining the original escape codes?

Comment (Manishearth) [reply clippy_lints/src/byte_char_slice.rs:74]: Ah, i think it does. Could you add some test cases that have stuff like `\x00` and `\x01` as byte literals?

Comment (Alexendoo) [reply clippy_lints/src/byte_char_slice.rs:74]: Might need to handle `b'"'` specifically since that'll need a `\`  in a byte string

Comment (Alexendoo) [reply clippy_lints/src/byte_char_slice.rs:74]: We may need to handle `b'"'` specifically since it'll need a `\` in a byte string

Comment (llogiq) [reply clippy_lints/src/byte_char_slice.rs:74]: Conversely, we may remove the backslash before single quotes as in `'\''`.

File: clippy_lints/src/byte_char_slice.rs
Comment (TheNeikos) [reply clippy_lints/src/byte_char_slice.rs]: Done!

File: clippy_lints/src/byte_char_slice.rs
Comment (TheNeikos) [reply clippy_lints/src/byte_char_slice.rs]: I've added all three of these cases!

File: tests/ui/byte_char_slice.stderr
Comment (Manishearth) [inline tests/ui/byte_char_slice.stderr:24]: issue: wrong kind of slash!

**Review by Manishearth — CHANGES_REQUESTED** [review]

File: tests/ui/byte_char_slice.stderr
Comment (llogiq) [reply tests/ui/byte_char_slice.stderr:24]: Could that be a compiletest artifact? The `.fixed` version looks OK.

Comment (Manishearth) [reply tests/ui/byte_char_slice.stderr:24]: hmmmmm yeah probably!

feels like a bug

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
```
---- src/byte_char_slice.rs - byte_char_slice::BYTE_CHAR_SLICE (line 21) stdout ----
error[E0308]: mismatched types
 --> src/byte_char_slice.rs:22:1
  |
2 | fn main() { #[allow(non_snake_case)] fn _doctest_main_src_byte_char_slice_rs_21_0() {
  |                                                                                     - help: try adding a return type: `-> &'static [u8; 5]`
3 | b"Hello"
  | ^^^^^^^^ expected `()`, found `&[u8; 5]`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
Couldn't compile the test.
---- src/byte_char_slice.rs - byte_char_slice::BYTE_CHAR_SLICE (line 17) stdout ----
error[E0308]: mismatched types
 --> src/byte_char_slice.rs:18:1
  |
2 | fn main() { #[allow(non_snake_case)] fn _doctest_main_src_byte_char_slice_rs_17_0() {
  |                                                                                     - help: try adding a return type: `-> &[u8; 5]`
3 | &[b'H', b'e', b'l', b'l', b'o']
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `()`, found `&[u8; 5]`

error: aborting due to previous error

```

File: tests/ui/byte_char_slice.stderr
Comment (TheNeikos) [reply tests/ui/byte_char_slice.stderr:24]: Yeah I was wondering about this as well, but it does the right thing anyway, so I don't know what I could do

Comment (llogiq) [reply tests/ui/byte_char_slice.stderr:24]: Having reviewed compiletest code, I'm now pretty sure this is a slash normalization to ease dealing with paths across windows/unices.

**Comment by llogiq** [comment]
@TheNeikos I think we can merge this after a rebase.

**Comment by xFrednet** [comment]
Hey @TheNeikos this is a ping from triage. Would you mind rebasing this PR?

@rustbot author

**Comment by xFrednet** [comment]
I've just rebased, since that seamed to be the last thing holding this back. I'll create an FCP for this.

**Comment by xFrednet** [comment]
I'd say the FCP is done. The only feedback was to rename the lint, which I've done.

@Manishearth Can we r+ it, or do you want to check something else?

FCP: https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/FCP.3A.20.60byte_char_slice.60.20rust-clippy.2310155/near/445939887

**Comment by Manishearth** [comment]
Yeah that sounds good!

@bors r+

**Comment by danielocfb** [comment]
I think this lint has a [false positive](https://github.com/libbpf/blazesym/actions/runs/10724198163/job/29739299322#step:4:576)
```
error: can be more succinctly written as a byte str
   --> src/symbolize/perf_map.rs:381:58
    |
381 |         let () = child.stdin.as_ref().unwrap().write_all(&[b'\n']).unwrap();
    |                                                          ^^^^^^^^ help: try: `b"\n"`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#byte_char_slices
    = note: `-D clippy::byte-char-slices` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::byte_char_slices)]`
```
The suggestion seems blatantly wrong, as the code doesn't even compile if I follow it:
```
error[E0308]: mismatched types
   --> src/symbolize/perf_map.rs:382:60
    |
382 |         let () = child.stdin.as_ref().unwrap().write_all(&[b"\n"]).unwrap();
    |                                                            ^^^^^ expected `u8`, found `&[u8; 1]`
```

**Comment by y21** [comment]
@danielocfb The arrows in the diagnostic cover the whole function argument, so clippy is suggesting `.write_all(b"\n")`, which should compile.

**Comment by danielocfb** [comment]
Ah my bad. Sorry for the noise then.

**Pull request merged**
